### PR TITLE
Add plural resources

### PR DESCRIPTION
**Pull request by skpaterson**
Includes resources required for upcoming blog post.

### CODE REVIEW

File: libraries/google_compute_firewalls.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      @display_name = opts[:name]
+      @project = opts[:project]
+    end
+
+    # FilterTable setup
+    filter_table_config = FilterTable.create
+    filter_table_config.add_accessor(:where)
+    filter_table_config.add_accessor(:entries)
+    filter_table_config.add(:exist?) { |filter_table| !filter_table.entries.empty? }
+    filter_table_config.add(:count) { |filter_table| filter_table.entries.count }
```
Comment (miah) [inline libraries/google_compute_firewalls.rb]: With https://github.com/inspec/inspec/pull/3104 we don't need to add `where`, `entries`, `exist?` or `count` properties to filter tables, they are now automatic.

File: libraries/google_compute_zones.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      @display_name = opts[:name]
+      @project = opts[:project]
+    end
+
+    # FilterTable setup
+    filter_table_config = FilterTable.create
+    filter_table_config.add_accessor(:where)
+    filter_table_config.add_accessor(:entries)
+    filter_table_config.add(:exist?) { |filter_table| !filter_table.entries.empty? }
+    filter_table_config.add(:count) { |filter_table| filter_table.entries.count }
```
Comment (miah) [inline libraries/google_compute_zones.rb]: As above.

File: docs/resources/google_compute_firewalls.md.erb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+<br>
+
+## Examples
+
+The following examples show how to use this InSpec audit resource.
+
+### Test that there are no more than a specified number of firewalls available for the project
+
+    describe google_compute_firewalls(project: 'chef-inspec-gcp') do
+      its('entries.count') { should be <= 100}
```
Comment (clintoncwolfe) [inline docs/resources/google_compute_firewalls.md.erb]: You can just use `its('count') ...` here.

File: docs/resources/google_compute_firewalls.md.erb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    end
+
+### Test the exact number of firewalls in the project
+
+    describe google_compute_firewalls(project: 'chef-inspec-gcp') do
+      its('firewall_ids.count') { should cmp 8 }
```
Comment (clintoncwolfe) [inline docs/resources/google_compute_firewalls.md.erb]: I'm not clear on the difference between this example and the previous one.  If the resource matches 5 firewalls, are you not expected to have 5 firewall IDs?

File: docs/resources/google_compute_firewalls.md.erb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      its('firewall_names') { should include "my-app-firewall-rule" }
+    end
+
+### Test that a particular named rule does not exist
+
+    describe google_compute_firewalls(project: 'chef-inspec-gcp') do
+      its('firewall_names') { should_not include "default-allow-ssh" }
+    end
+
+### Test there are no firewalls for the "INGRESS" direction
```
Comment (clintoncwolfe) [inline docs/resources/google_compute_firewalls.md.erb]: Hrm, this is usually expressed like this:
```
describe google_compute_firewalls(project: 'chef-inspec-gcp').where(firewall_direction: 'INGRESS') do
  it { should_not exist }
end
```

File: docs/resources/google_compute_firewalls.md.erb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+    describe google_compute_firewalls(project: 'chef-inspec-gcp') do
+      its('firewall_directions') { should_not include "INGRESS" }
+    end
+
+<br>
+
+## Filter Criteria
+
+This resource currently does not support any filter criteria; it will always fetch all available firewalls.
```
Comment (clintoncwolfe) [inline docs/resources/google_compute_firewalls.md.erb]: Actually, you have `firewall_id`, `firewall_name`, and `firewall_direction`.  You can use any of them with `where`, as a block or as a method.

Please read https://github.com/inspec/inspec/blob/master/docs/dev/filtertable-usage.md

File: docs/resources/google_compute_firewalls.md.erb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+## Properties
+
+*  `firewall_id`, `firewall_name`, `firewall_directions`
```
Comment (clintoncwolfe) [inline docs/resources/google_compute_firewalls.md.erb]: The properties are all the plural forms - so `firewall_ids`, `firewall_names`.  Typical would be to explicitly document each, describing what they are (likely arrays of strings, but what do the Strings mean?)

File: docs/resources/google_compute_instance.md.erb [hunk 8]
```diff
@@ -75,11 +75,17 @@ The following examples show how to use this InSpec audit resource.
       its('first_network_interface_type'){ should eq "one_to_one_nat" }
     end
 
+### Test that a particular compute instance label key is present
+
+    describe google_compute_instance(project: 'chef-inspec-gcp',  zone: 'us-east1-b', name: 'inspec-test-vm') do
```
Comment (clintoncwolfe) [inline docs/resources/google_compute_instance.md.erb]: This seems to be an unrelated change to the PR

File: docs/resources/google_compute_instance.md.erb [hunk 9]
```diff
@@ -75,11 +75,17 @@ The following examples show how to use this InSpec audit resource.
...
+      its('labels_keys') { should include 'my_favourite_label' }
+    end
+
 <br>
 
 ## Properties
 
-*  `cpu_platform`, `creation_timestamp`, `deletion_protection`, `disks`, `id`, `kind`, `label_fingerprint`, `machine_type`, `metadata`, `name`, `network_interfaces`, `scheduling`, `start_restricted`, `status`, `tags`, `zone`
+*  `cpu_platform`, `creation_timestamp`, `deletion_protection`, `disks`, `id`, `kind`, `label_fingerprint`, `machine_type`, `metadata`, `name`, `network_interfaces`, `scheduling`, `start_restricted`, `status`, `tags`, `zone`, `labels_keys`, `labels_values`
```
Comment (clintoncwolfe) [inline docs/resources/google_compute_instance.md.erb]: We should talk about a tagging / label API.  Azure does it with dynamic methods; AWS doesn't really do it at all yet.  We should do it in a similar way, and I think a custom matcher might make more sense than a pair of properties.

File: docs/resources/google_compute_zone.md.erb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+### Test that a GCP compute zone does not exist
+
+    describe google_compute_zone(project: 'chef-inspec-gcp',  zone: 'us-east1-b') do
+      it { should_not exist }
+    end
+
+### Test that a GCP compute zone is in the expected state
+
+    describe google_compute_zone(project: 'chef-inspec-gcp',  zone: 'us-east1-b') do
```
Comment (clintoncwolfe) [inline docs/resources/google_compute_zone.md.erb]: This seems like a great chance for a bit of UX sugar.  If you defined a method `def up? ...`, you'd have a `be_up` matcher, and you can say 
```
describe google_compute_zone(...) do
  it { should be_up }
end
```

File: docs/resources/google_compute_zones.md.erb [hunk 11]
```diff
@@ -0,0 +1,75 @@
+---
+title: About the google_compute_zones Resource
+platform: gcp
+---
+
+# google\_compute\_zones
+
+Use the `google_compute_zones` InSpec audit resource to test properties of all GCP compute zones for a project in a particular zone.
```
Comment (clintoncwolfe) [inline docs/resources/google_compute_zones.md.erb]: ...all or a filtered group of ...

File: docs/resources/google_compute_zones.md.erb [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+<br>
+
+## Examples
+
+The following examples show how to use this InSpec audit resource.
+
+### Test that there are no more than a specified number of zones available for the project
+
+    describe google_compute_zones(project: 'chef-inspec-gcp') do
```
Comment (clintoncwolfe) [inline docs/resources/google_compute_zones.md.erb]: So, similar comments as on `firewalls`.  Use `count` instead of `entries.count`.  You do actually have filter criteria.  Your properties are plural.

File: libraries/google_compute_firewalls.rb [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    filter_table_config.add(:firewall_ids, field: :firewall_id)
+    filter_table_config.add(:firewall_names, field: :firewall_name)
+    filter_table_config.add(:firewall_directions, field: :firewall_direction)
+    filter_table_config.add(:colors, field: :color, type: :simple)
```
Comment (clintoncwolfe) [inline libraries/google_compute_firewalls.rb]: Colors appears to be copypasta.  Check your other plurals as well.

File: libraries/google_compute_firewalls.rb [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    filter_table_config.connect(self, :fetch_data)
+
+    def fetch_data
+      firewall_rows = []
+      catch_gcp_errors do
+        @firewalls = @gcp.gcp_compute_client.list_firewalls(@project)
```
Comment (clintoncwolfe) [inline libraries/google_compute_firewalls.rb]: Do you need to paginate?

File: test/integration/verify/controls/google_compute_firewalls.rb [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+control 'gcp-firewalls-1.0' do
+
+  impact 1.0
+  title 'Ensure firewalls have the correct properties in bulk'
+
+  describe google_compute_firewalls(project: gcp_project_id) do
+    it { should exist }
+    its('entries.count') { should be <= 100}
+    # assume this is a development setup for a moment
+    its('firewall_names') { should include "default-allow-ssh" }
```
Comment (clintoncwolfe) [inline test/integration/verify/controls/google_compute_firewalls.rb]: You should also test firewall_ids - any property you expose.  To obtain a value to check for, you can use the "output the ID in terraform, translate to YAML, read as an attribute" approach.  AWS does this.  Other methods also work.

File: libraries/google_compute_firewalls.rb (hunk 1 above)
Comment (skpaterson) [reply libraries/google_compute_firewalls.rb]: Happy to do this but worth noting all profiles using inspec-gcp will now have to update the minimum InSpec version because of https://github.com/inspec/inspec/issues/3066

File: docs/resources/google_compute_zone.md.erb [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+      its('name') { should match 'us-east1-b' }
+    end
+
+<br>
+
+## Examples
+
+The following examples show how to use this InSpec audit resource.
+
+### Test that a GCP compute zone does not exist
```
Comment (erjohnso) [inline docs/resources/google_compute_zone.md.erb]: Suggest flipping logic to test that a specified zone *does* exist.

File: libraries/google_compute_firewalls.rb (hunk 14 above)
Comment (erjohnso) [reply libraries/google_compute_firewalls.rb]: The API supports pagination [https://cloud.google.com/compute/docs/reference/rest/v1/firewalls/list] but the default is 500 so it's probably fine to skip that for now.

File: docs/resources/google_compute_firewalls.md.erb (hunk 3 above)
Comment (skpaterson) [reply docs/resources/google_compute_firewalls.md.erb]: sure, will update

File: docs/resources/google_compute_firewalls.md.erb (hunk 4 above)
Comment (skpaterson) [reply docs/resources/google_compute_firewalls.md.erb]: agreed, will remove

File: docs/resources/google_compute_firewalls.md.erb (hunk 5 above)
Comment (skpaterson) [reply docs/resources/google_compute_firewalls.md.erb]: thanks, will update the example

File: docs/resources/google_compute_firewalls.md.erb (hunk 6 above)
Comment (skpaterson) [reply docs/resources/google_compute_firewalls.md.erb]: thanks, likely another copy/paste oversight, will update accordingly

File: docs/resources/google_compute_firewalls.md.erb (hunk 7 above)
Comment (skpaterson) [reply docs/resources/google_compute_firewalls.md.erb]: sure

File: docs/resources/google_compute_instance.md.erb (hunk 8 above)
Comment (skpaterson) [reply docs/resources/google_compute_instance.md.erb]: was related to the upcoming blog post example that uses the plural resources

File: docs/resources/google_compute_instance.md.erb (hunk 9 above)
Comment (skpaterson) [reply docs/resources/google_compute_instance.md.erb]: that would be great - let's start with what's there and iterate

File: docs/resources/google_compute_zone.md.erb (hunk 10 above)
Comment (skpaterson) [reply docs/resources/google_compute_zone.md.erb]: will add this - much more readable

File: docs/resources/google_compute_zones.md.erb (hunk 11 above)
Comment (skpaterson) [reply docs/resources/google_compute_zones.md.erb]: will add

File: docs/resources/google_compute_zones.md.erb (hunk 12 above)
Comment (skpaterson) [reply docs/resources/google_compute_zones.md.erb]: agreed

File: libraries/google_compute_firewalls.rb (hunk 13 above)
Comment (skpaterson) [reply libraries/google_compute_firewalls.rb]: yup, removed

File: libraries/google_compute_firewalls.rb (hunk 14 above)
Comment (skpaterson) [reply libraries/google_compute_firewalls.rb]: defaults to 500 but will add pagination

File: docs/resources/google_compute_zone.md.erb (hunk 16 above)
Comment (skpaterson) [reply docs/resources/google_compute_zone.md.erb]: sure can do - thanks

**Comment by skpaterson** [comment]
Thanks @erjohnso for your comments - created https://github.com/inspec/inspec-gcp/issues/27 to track the firewall rule improvements.

File: test/integration/verify/controls/google_compute_firewalls.rb (hunk 15 above)
Comment (skpaterson) [reply test/integration/verify/controls/google_compute_firewalls.rb]: added tests for the plural resource id parameters

**Comment by skpaterson** [comment]
Thanks for the feedback, hopefully the comments should all now be addressed (famous last words...)

File: docs/resources/google_compute_firewalls.md.erb [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+      it { should_not exist }
+    end
+
+<br>
+
+## Filter Criteria
+
+This resource supports the following filter criteria:  `firewall_id`; `firewall_name`; and `firewall_direction`. Any of these may be used with `where`, as a block or as a method.
+
+## Properties
```
Comment (clintoncwolfe) [inline docs/resources/google_compute_firewalls.md.erb]: Let's go ahead and ship this, but typically we include one example per property.  Users links straight to the property, then copy and paste the example.

File: docs/resources/google_compute_instance.md.erb (hunk 9 above)
Comment (clintoncwolfe) [reply docs/resources/google_compute_instance.md.erb]: I'd be reluctant to add anything you want to remove later.  It's very hard to deprecate once thing are in the wild.  Since this isn't in core, let's ship it, but keep in mind if we have doubts about something, better to not expose it.

Comment (clintoncwolfe) [reply docs/resources/google_compute_instance.md.erb]: See https://github.com/inspec/inspec/issues/3139

File: docs/resources/google_compute_instances.md.erb [hunk 18]
```diff
@@ -57,11 +57,12 @@ The following examples show how to use this InSpec audit resource.
 
 ## Filter Criteria
 
-This resource currently does not support any filter criteria; it will always fetch all instances in the zone.
+This resource supports the following filter criteria:  `instance_id` and `instance_name`. Either of these may be used with `where`, as a block or as a method.
```
Comment (clintoncwolfe) [inline docs/resources/google_compute_instances.md.erb]: If you need it for a blog post, let's ship it, but please make a tech debt issue - people fiind `where` very confusing, and need examples.

File: libraries/google_compute_firewalls.rb [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+      super(opts)
+      @display_name = opts[:name]
+      @project = opts[:project]
+    end
+
+    # FilterTable setup
+    filter_table_config = FilterTable.create
+    filter_table_config.add(:firewall_ids, field: :firewall_id)
+    filter_table_config.add(:firewall_names, field: :firewall_name)
+    filter_table_config.add(:firewall_directions, field: :firewall_direction)
```
Comment (clintoncwolfe) [inline libraries/google_compute_firewalls.rb]: This is good as-is, but you might consider adding `style: :simple` - having this de-duplicated might be useful (or at least having individual entries for each rule isn't very useful).
